packages/block-serialization-default-parser/parser.php

```diff
@@ -0,0 +1,266 @@
+                $stack_top->block->innerHTML .= substr( $this->document, $stack_top->prev_offset, $start_offset - $stack_top->prev_offset );
+                $stack_top->prev_offset = $start_offset + $token_length;
+
+                $this->add_inner_block(
+                    $stack_top->block,
+                    $stack_top->token_start,
+                    $stack_top->token_length,
+                    $start_offset + $token_length
+                );
+                $this->offset = $start_offset + $token_length;
+                return true;
+
+            default:
+                // This is an error
+                $this->add_freeform();
+                return false;
+        }
+    }
+
+    function next_token() {
+        $matches = null;
+
+        /*
+         * aye the magic
+         * we're using a single RegExp to tokenize the block comment delimiters
+         * we're also using a trick here because the only difference between a
+         * block opener and a block closer is the leading `/` before `wp:` (and
+         * a closer has no attributes). we can trap them both and process the
+         * match back in PHP to see which one it was.
+         */
+        $has_match = preg_match(
+            '/<!--\s+(?<closer>\/)?wp:(?<namespace>[a-z][a-z0-9_-]*\/)?(?<name>[a-z][a-z0-9_-]*)\s+(?<attrs>{(?:(?!}\s+-->).)+?}\s+)?(?<void>\/)?-->/s',
+            $this->document,
+            $matches,
+            PREG_OFFSET_CAPTURE,
+            $this->offset
+        );
+
+        // we have no more tokens
+        if ( 0 === $has_match ) {
+            return array( 'no-more-tokens' );
+        }
+
+        list( $match, $started_at ) = $matches[ 0 ];
+
+        $length     = strlen( $match );
+        $is_closer  = isset( $matches[ 'closer' ] ) && -1 !== $matches[ 'closer' ][ 1 ];
+        $is_void    = isset( $matches[ 'void' ] ) && -1 !== $matches[ 'void' ][ 1 ];
+        $namespace  = $matches[ 'namespace' ];
+        $namespace  = ( isset( $namespace ) && -1 !== $namespace[ 1 ] ) ? $namespace[ 0 ] : 'core/';
+        $name       = $namespace . $matches[ 'name' ][ 0 ];
+        $has_attrs  = isset( $matches[ 'attrs' ] ) && -1 !== $matches[ 'attrs' ][ 1 ];
+        $attrs      = $has_attrs ? json_decode( $matches[ 'attrs' ][ 0 ] ) : null;
+
+        /*
+         * This state isn't allowed
+         * This is an error
+         */
+        if ( $is_closer && ( $is_void || $has_attrs ) ) {
+            // we can ignore them since they don't hurt anything
+        }
+
+        if ( $is_void ) {
+            return array( 'void-block', $name, $attrs, $started_at, $length );
+        }
+
+        if ( $is_closer ) {
+            return array( 'block-closer', $name, null, $started_at, $length );
+        }
+
+        return array( 'block-opener', $name, $attrs, $started_at, $length );
+    }
+
+    function add_freeform( $length = null ) {
+        $length = $length ?: strlen( $this->document ) - $this->offset;
```

Comment:
The `?:` ternary shorthand was only [introduced](https://secure.php.net/manual/en/language.operators.comparison.php) in PHP 5.3, while we need to support 5.2.

Reply:
Addressed in 80279bb99.